### ISSUE TITLE
Same message is consumed multiple times by consumers running on multiple instances

### ISSUE THREAD

**Issue by Siddharth-Somani** [spring-cloud/spring-cloud-stream-binder-aws-kinesis#148]
Duplicate messages are consumed by multiple consumers running on different instances. We have a setup with 6 pods and 8 kinesis shards.
These are the spring properties in my application - 

```
spring.cloud.stream.bindings.input.destination=kinesis_stream
spring.cloud.stream.bindings.input.group=kinesis_stream-group
spring.cloud.stream.bindings.input.content-type=application/json
spring.cloud.stream.bindings.enrichOutput.destination= kinesis_stream
spring.cloud.stream.bindings.enrichOutput.content-type= application/json
spring.cloud.stream.bindings.enrichOutput.producer.headerMode = none
spring.cloud.stream.kinesis.bindings.enrichOutput.producer.sync = true
```

A few observations - 
1. Duplicate messages are being consumed only after changing producer to synchronous. But only single message is being produced by producer regardless of sync or async
2. In DynamoDB `SpringIntegrationLockRegistry` table, each shard has an exclusive instance attached which is the expected behavior. 
3. Duplicate messages are being consumed multiple times by the same consumer instance as well as different consumer instances

Libraries used - 

```
<spring-cloud-stream-binder-kinesis.version>2.0.1.RELEASE</spring-cloud-stream-binder-kinesis.version>
 <spring-cloud-stream.version>3.0.1.RELEASE</spring-cloud-stream.version>
 <spring-integration-aws.version>2.3.4.RELEASE</spring-integration-aws.version>
```

Please let me know if any other details are required

**Comment by artembilan** [comment]
Not sure what is going on, but here are several possibilities:

1. You really need to be sure that `group` is the same on all the instances.
2. Check, please, what you have so far in the `SpringIntegrationMetadataStore` DynamoDb table. It has to contain some updated checkpoints for those consumed shards.
3. You really need to be sure that we talk about the same Kinesis Record, not the same data in them. You producer may fluctuate and therefore send several records into the stream, but data is the same. 

So, "same message" need to be confirmed as really **same Kinesis record**. You can turn on a DEBUG logging level for `org.springframework.integration` category to track the data and headers you consume to really see that the same offset from the same shard is processed in different or even the same consumer. More over you confirm that you have an exclusive locks for those shards in the appropriate `SpringIntegrationLockRegistry` table. Therefore only one consumer take records from one shard at a time. The same record can't go to different shard and therefore another consumer.

**Comment by Siddharth-Somani** [comment]
1. Yes, the group is same on all the instances
2. `SpringIntegrationMetadataStore` has some updated checkpoints for each shard.
3. Confirmed that its the same kinesis record, not just data. It has same shard, same sequence number and same arrival timestamp.
4. Also confirmed that each shard has a different owner (the host) in the `SpringIntegrationLockRegistry` table.

Yet, multiple consumers are consuming the same kinesis record. Point 4 is having no effect.

**Comment by artembilan** [comment]
> It has same shard, same sequence number and same arrival timestamp.

Sounds like a *retry*, but it is possible only on a single consumer and only within that consumer. There is nothing like "re-delivery" in Kinesis.

With your point 4 it is fully not clear how another instance has access to the same shard to be able to consume the same record...

**Comment by Siddharth-Somani** [comment]
Exactly my issue with point 4. DynamoDB entry suggests that each shard has only one owner. Still multiple instances are consuming from the same shard. Is there something else that I can add in the properties ?

**Comment by artembilan** [comment]
May it be the fact that your consumers are in different regions so they have their own `SpringIntegrationLockRegistry` and `SpringIntegrationMetadataStore` tables? Therefore they are treated as different groups... Kinesis doesn't have any "consumer group" logic by itself - this is done internally in the Spring Integration AWS and based on those mention shared metadata and lock registry stores.

I'm fully not sure how your "Duplicate messages are being consumed only after changing producer to synchronous" may makes sense... The difference is only if we wait for the send completion or not.

**Comment by Siddharth-Somani** [comment]
Our application is hosted only in one AWS region, so all consumers are from the same region. 
Yes, I know about the consumer group logic.

**Comment by artembilan** [comment]
OK. So, make sure that all your instances are looking really into the same `SpringIntegrationLockRegistry`. Otherwise, if they do, what you are claiming here just does not make sense. When all the consumers in the group deal with the same stream, only one instance can consume from a single shard. Just because there is a shared lock on that shard with the group. Plus your point about `sync` mode is fully out of scope: I definitely don't see how producer can give any effect here, since you confirm that duplication is on the record level in shard not with the data.

**Comment by artembilan** [comment]
Anyway: try your solution with the `<spring-cloud-stream-binder-kinesis.version>2.0.3.RELEASE</spring-cloud-stream-binder-kinesis.version>`. I don't think you need the rest of the mentioned dependencies.

**Comment by artembilan** [comment]
I found another logic over there:
1. `KinesisMessageDrivenChannelAdapter`
```
if (this.streams == null) {
	if (this.lockRegistry != null) {
		logger.warn("The LockRegistry is ignored when explicit shards configuration is used.");
	}
	this.lockRegistry = null;
}
```
2. `KinesisMessageChannelBinder`
```
if (properties.getInstanceCount() > 1) {
			shardOffsets = new HashSet<>();
			KinesisConsumerDestination kinesisConsumerDestination = (KinesisConsumerDestination) destination;
			List<Shard> shards = kinesisConsumerDestination.getShards();
			for (int i = 0; i < shards.size(); i++) {
				// divide shards across instances
				if ((i % properties.getInstanceCount()) == properties.getInstanceIndex()) {
					KinesisShardOffset shardOffset = new KinesisShardOffset(
							kinesisShardOffset);
					shardOffset.setStream(destination.getName());
					shardOffset.setShard(shards.get(i).getShardId());
					shardOffsets.add(shardOffset);
				}
			}
		}

		KinesisMessageDrivenChannelAdapter adapter;

		...

		if (CollectionUtils.isEmpty(shardOffsets)) {
			adapter = new KinesisMessageDrivenChannelAdapter(amazonKinesisClient, destination.getName());
		}
		else {
			adapter = new KinesisMessageDrivenChannelAdapter(amazonKinesisClient,
					shardOffsets.toArray(new KinesisShardOffset[0]));
		}
```

So, if there are `instanceCount` & `instanceIndex` Spring Cloud Stream consumer properties, then consumer group is fully ignored and target consumer deals only with those assigned shards. From here I can only assume that one of your consumers is good and operates with consumer group properly, but the rest are adjusted somehow with those `instance` properties and more over - with the same values(!!). Therefore the same shard can be consumed in different places.

**Comment by Siddharth-Somani** [comment]
You can ignore my comment on sync/async part. It was a confusion on my side.

Anyways, I tried upgrading `spring-cloud-stream-binder-kinesis.version` to `2.0.3.RELEASE`. Still no effect. Multiple consumer instances are still consuming the same record. I also tried to change checkpointing to record instead of batch but that also didn't work.

I have not explicitly defined `instanceCount` and `instanceIndex`. The only properties that I have is in the question above. Do I explicitly need to define instanceIndex in my properties file? 

Meanwhile I am changing logging level to DEBUG to see if I find something.

**Comment by artembilan** [comment]
Well, those properties are set by cloud environment. 
See Spring Cloud Stream:
https://docs.spring.io/spring-cloud-stream/docs/3.0.10.RELEASE/reference/html/spring-cloud-stream.html#spring-cloud-stream-overview-instance-index-instance-count

It really would be great if you can debug several instances of your application to be sure what is going on.
But for me it sounds like they all have been assigned with the same shard, not stream, so according the logic I provided before, the lock is ignored and several instances are able to consume from the same shard even if they belong to the same consumer group. Such a logic is fully aligned with Apache Kafka manual partition assignment. Anyway: we need to confirm that first of all. I don't see how the same shard could be consumed with several instances if they are in the same group and configuration is based only on the stream - no manual shards assigning.

**Comment by Siddharth-Somani** [comment]
Thanks @artembilan. Looking at this comment of yours in another thread, I did not add `instanceCount` and `instanceIndex` in my application properties - https://github.com/spring-cloud/spring-cloud-stream/issues/1342#issuecomment-378701598. The only way I am aware of to debug this problem is to look into the tables in DynamoDB which seems to have no issue.
Do you have any suggestions on how I can debug this on a instance level and check what default values are being set for the two properties? Has anyone else raised a similar issue before like this?

**Comment by artembilan** [comment]
No, no one raised such a concern before. I mean duplicated records when all the env seems OK.

Only the way to analyze I see like a DEBUG logging level for the whole `org.springframework` category.
So, you probably will see in logs all the configuration properties.
Plus you will see the message flow and some info for records and shards they've been consumed from.
There also has to be that "The LockRegistry is ignored when explicit shards configuration is used" warning because we definitely inject `lockRegistry` from the Binder configuration.

If that is not, then your instances look into different DynamoDBs somehow. Or `LockRegistry` is different for them...

**Comment by Siddharth-Somani** [comment]
I have enabled debug logs for org.springframework, cannot see the log `The LockRegistry is ignored when explicit shards configuration is used.`. Which log tells about the default instanceIndex and count? I am unable to see them. Any other specific debug logs that I need to check?

**Comment by sabbyanandan** [comment]
Hi, @Siddharth-Somani. I am jumping in to understand your environment. It'd be good if you can answer the following questions.

1) It'd be nice to review the app and its properties that you're overriding. If you can sensitive stuff in the app, please remove and share the bare bones of the app in a GitHub repo.
2) How are you deploying the app? Specifically, where? A clear description of the steps involved in the deployment would be useful.
3) Extension to the previous question. Would it be possible to explain the deployment topology? Since there are multiple instances running in a distributed mode, how are each app provisioned on the given platform? (K8s? CF? VMs?)
4) Lastly, if there's anything else you can share to quickly reproduce this behavior, we could review it on our side.

**Comment by artembilan** [comment]
As far as I know there is no way to log the application properties.
Or you should write your own logger or just take a look into an Actuator `/env` endpoint: https://stackoverflow.com/questions/24448947/how-to-log-the-active-configuration-in-a-spring-boot-application.

May be all your instances are started in different AWS accounts?..

Anyway: as @sabbyanandan is asking it would be great to have some simple project from your and description how to reproduce.
It is just not visible from our side how that would be possible if all the instances in the same account, same group and have proper shard locking report...

Thank you for understanding!

**Comment by Siddharth-Somani** [comment]
1. I cannot share the code due to my company's policy. The producer and consumer are exactly the same as what is described in the spring cloud stream documentation.
Producer - 
`@EnableBinding(Source.class)
public class Producer {

    @Autowired
    private Source source;

    public void send(String data) {
        source.output().send(MessageBuilder.withPayload(data).build());
    }
}`

Consumer - 
`@EnableBinding(Sink.class)
public class Consumer {

    @StreamListener(Sink.INPUT)
    public void consume(Message<String> message) {
        log.info(message.getPayload());
    }`

The properties are mentioned in the original question above.

2. We have deployed our application in dockerized containers using K8s. We are running currently on 4 pods.

3. One important thing to mention - The code is deployed in a different aws account but the other resources such as kinesis and dynamo are in a different aws account. 
I am using STS assume role to initialize the AWS resources config. These are the only configs - 
`@Bean
    public AmazonDynamoDB amazonDynamoDB() {
        STSAssumeRoleSessionCredentialsProvider provider = new STSAssumeRoleSessionCredentialsProvider.Builder(assumeRole, sessionName).build();
        return AmazonDynamoDBAsyncClientBuilder.standard()
                .withCredentials(provider)
                .withRegion(awsRegion)
                .build();
    }

    @Bean
    public AmazonKinesisAsync amazonKinesis() {
        STSAssumeRoleSessionCredentialsProvider provider = new STSAssumeRoleSessionCredentialsProvider.Builder(assumeRole, sessionName).build();
        return AmazonKinesisAsyncClientBuilder.standard()
                .withCredentials(provider)
                .withRegion(awsRegion)
                .build();
    }

    @Bean
    public KinesisProducer kinesisProducer() {
        STSAssumeRoleSessionCredentialsProvider provider = new STSAssumeRoleSessionCredentialsProvider.Builder(assumeRole, sessionName).build();
        KinesisProducerConfiguration config = new KinesisProducerConfiguration()
                .setRegion(awsRegion)
                .setAggregationEnabled(false)
                .setCredentialsProvider(provider);
        return new KinesisProducer(config);
    }`

Let me know if I need to change something over here or remove any config.

4. Please find the image of the SpringIntegrationLockRegistry table attached. As mentioned above - these are in a different aws account as compared to the original code. But each shard is assigned only one node as you can see. 
<img width="1237" alt="Screenshot 1399-09-19 at 1 01 55 AM" src="https://user-images.githubusercontent.com/18643496/101532219-3b81e180-39ba-11eb-8390-0ad78365003e.png">

**Comment by artembilan** [comment]
You say `KinesisProducer` in your config. So looks like you use a KPL to produce into the Kinesis stream.
Do you use KCL on the consumer side? If not, the regular `KinesisMessageDrivenChannelAdapter` is not compatible with the data produced by the KPL. Consider to use a `kplKclEnabled=true` for that purpose.

**Comment by Siddharth-Somani** [comment]
I don't think `KinesisProducer` KPL is being used since we are pushing to kinesis synchronously and have set the `producer.sync` as true. The reason for defining `AmazonKinesisAsync` and `AmazonDynamoDB` explicitly is that we get a permission error (Due to it being in a different aws account). What are the configs we need according to you in order to have the behaviour that I mentioned?

**Comment by artembilan** [comment]
Any chances that you can try your solution with the `kplKclEnabled=true` all the way from producer to consumer?

Be aware that you probably also need to expose custom bean for the `AmazonCloudWatchAsync` as well, since it is used by KCL.
Also what I need to revise your `KinesisProducerConfiguration` as a bean since we don't care about `KinesisProducer` by itself, but more about a `KinesisProducerConfiguration` : 

```
@Bean
	@ConditionalOnMissingBean
	@ConditionalOnProperty(name = "spring.cloud.stream.kinesis.binder.kpl-kcl-enabled")
	public KinesisProducerConfiguration kinesisProducerConfiguration() {
		KinesisProducerConfiguration kinesisProducerConfiguration = new KinesisProducerConfiguration();
		kinesisProducerConfiguration.setCredentialsProvider(this.awsCredentialsProvider);
		kinesisProducerConfiguration.setRegion(this.region);
		return kinesisProducerConfiguration;
	}
```

On the other hand you may don't need those custom beans at all, but just a `AWSCredentialsProvider` for your STS.
See for example:

```
	@Bean
	@ConditionalOnMissingBean
	public AmazonKinesisAsync amazonKinesis() {
		return AmazonKinesisAsyncClientBuilder.standard()
				.withCredentials(this.awsCredentialsProvider)
				.withRegion(this.region)
				.build();
	}
```

Also, to be honest, you may try to change your `amazonDynamoDB` return type to the `AmazonDynamoDBAsync`, because bean condition works exactly on bean definition type:
```
	@Bean
	@ConditionalOnMissingBean
	public AmazonDynamoDBAsync dynamoDB() {
		if (this.hasInputs) {
			return AmazonDynamoDBAsyncClientBuilder.standard()
					.withCredentials(this.awsCredentialsProvider)
					.withRegion(this.region)
					.build();
		}
		else {
			return null;
		}
	}
```

**Comment by Siddharth-Somani** [comment]
Thanks, let me try this. Just a few more questions. Is it good to use KPL for synchronous use case? I think I need to provide `KinesisClientLibConfiguration` config as well for KCL. `KinesisClientLibConfiguration` expects workerIds as well. How exactly can I create a Bean for `KinesisClientLibConfiguration` ?

**Comment by Siddharth-Somani** [comment]
I also did not understand this `On the other hand you may don't need those custom beans at all, but just a AWSCredentialsProvider for your STS.`

**Comment by artembilan** [comment]
> I also did not understand this

There is no difference in the `AmazonKinesisAsync` bean definition provided in the library and yours.
Since that out-of-the-box bean can be autowired with external `AWSCredentialsProvider`, I would expect from your just a bean definition for that your `STSAssumeRoleSessionCredentialsProvider` - and the framework will inject it into  those beans:

```
@Bean
AWSCredentialsProvider stsCredentialsProvider() {
    return new STSAssumeRoleSessionCredentialsProvider.Builder(assumeRole, sessionName).build();
}
```

Yes, you can provide a `KinesisClientLibConfiguration` bean and it is going to be injected into the `KclMessageDrivenChannelAdapter` internally.

Have no idea how KPL works in the sync mode. Looks like there is no such an option of it. Although `recordMaxBufferedTime` might do something for us...

**Comment by Siddharth-Somani** [comment]
Thanks for the help. So I don't think we can use Kpl kcl then. And let me try with just creating one bean for `AWSCredentialsProvider` and remove other beans. Will update here after testing.

**Comment by malvine** [comment]
Hi,
 I think I've seen the same behaviour @Siddharth-Somani  is talking about twice already. But I can't reproduce it easily. But will try again.
_So far:_
**My case is** : I have **one stream with 2 shards and one consumer (with 2 instances)**. This is only happening with spring cloud kinesis binder with kpl_kcl feature off. (with kpl&kcl on haven't had this behaviour so far).

As the consumers are greedy - one instance gets 2 shards (I can see that in the DynamoDB table).
But eventually, after playing with killing one instance - so  i can see that another instance gets all the shards and then bringing the instance up again etc. , eventually,  I can see that one instance steals all the shards (I can see that in DynamoDB and in the logs) but there is nothing on another instance like a log that it's lost shards or anything like that. Which I found strange.
If I send the record to the stream - both instances get the same record. (even though only one instance holds all the shards in DynamoDB table). Seems like there is still the phantom connection of that first instance to the stream.

The interesting thing is let's say : _instance `one` has phantom connection and instance `two` has normal one._
If I kill the instance `two` (with normal connection) - the records in the table disappear (also normal expected behaviour), but if I put a record in the stream - the instance `one` (that still holds invisible connection to the stream) is still getting it.
If I bring the instance `two` back up - the DynamoDB will have 2 records in the table pointing to both shards for that instance `two`. And if I send the record still both instances will get it.
Only killing instance `one` will bring everything back to normal.

As I said I accidentally hit this case couple of times, but every time it was unexpected. I will try to play a bit more, hopefully I will spot something how easier to reproduce this.

**Comment by Siddharth-Somani** [comment]
@artembilan Removed the other beans. Now I just have one bean `AWSCredentialsProvider`. It still does not solve the issue. Still seeing duplicates. All nodes are consuming the same message.

**Comment by artembilan** [comment]
Well, it's very interesting and confusing issue.
I wonder why no one reported it before: what could we change in the logic to break it this way...

First of all it would be great to see some logs from your application to confirm that the same record is consumed by different instances.

Second: be sure that all the instances really look into the same DynamoDB table and there is no account-related independent table, so it would look like fully different environment even if Kinesis stream is the same.

Third: it would be interesting if you can test your solution with some other `LockRegistry` implemention, e.g. `JdbcLockRegistry` or `RedisLockRegistry`.

Another option to consider: how about to downgrade `com.amazonaws:dynamodb-lock-client` to version `1.0.0`?

And the last resort would be, of course, a simple project to let us to play and reproduce on our side.

Thanks for understanding

**Comment by malvine** [comment]
I just spotted something very interesting. Not sure if it's the same case as the one I've seen before, but still worth to mention.

I had 1 producer and 1 consumer (with 2 instances running locally) and one stream with one shard.

I produced an event and one of the consumer's instances got it. let's say _instance1_ got it. _instance2_ got nothing. we are happy. all good.
Then my wi-fi connection dropped just for a minute. Once the wi-fi connection has been restored I published another message and now both my instances _instance1_ and _instance2_ got it (in the dynamoDB _instance1_ still is owner of the lease). After that I can send as many messages as I want all my instances two instances of the same consumer (group) will get them.

To get it back to normal (so only one instance gets the message from the shard), I needed to re-bounce the instances. Nothing else helped.

@artembilan agree this is all very confusing. I'll try to get some more info if I'll have more.

**Comment by artembilan** [comment]
Thank you for sharing that observation, @malvine !

Please, confirm that lock in the DB is set for the second instance after network failure not the first.
My idea is that we may not monitor that we still hold a lock, but just trying to rely on its heartbeat feature. 
Therefore after having the network restored we continue to consume from the shard not having a lock.
But since the second instance sees the lock as expired, it can obtain it and start consuming.

I'm investigating the code for such a flaw...

**Comment by malvine** [comment]
@artembilan based on the example described above - Yes, after the network failure - the _instance2_ is now the owner of the lease in the DynamoDB. But both instances are getting records.

If I bring the _instance2_  down all my DynamoDB records disappear, like there is no owner of the lease, but interestingly the _instance1_ is still able to consume the records.

And if I bring the _instance2_ back up, I can see again, the record in the DynamoDB of _instance2_ owning the lease. But still both instances get the same record, until I take _instance1_ down.

Hope that helps.

**Comment by Siddharth-Somani** [comment]
@artembilan Downgrading `com.amazonaws:dynamodb-lock-client` didn't help. I'll try the remaining things that you mentioned and get back to you. How exactly can I change the LockRegistry implementation?

Also, I changed my `4 consumer nodes and 8 Kinesis Shards` setup to `4 consumer nodes and 2 Kinesis Shards`. By doing this, only one node has been consuming from both the shards. Other nodes are not consuming the same message which is the expected behaviour that we want. Unfortunately, we cannot operate on 2 shards for long.

**Comment by artembilan** [comment]
@malvine ,

thank you for confirmation!
So, yes, the logic is broken but it was like the from day first.
I can describe the flaw like this:
1. We take all the shards available from the provided streams and try to lease then from the start. Therefore only one instance have an ownership of the shard even if other instances come later. That's happy pass and it is a problem.
We just hold locks and do nothing with them: the consumer logic from the "locked" shard is done in the separate thread.
2. Now the problematic situation. When we lose the network there is no heartbeat from the lock to keep it leased. Therefore after bringing back network some other instance may already lock because it is expired. At the same time nothing hold the consumer thread from keeping to poll data from the shard it has locked in the past. Even if no network yet, we still able to call Kinesis API from our loop.

That's how it is implemented at the moment.

I probably see the fix like check the lock state (or just `tryLock()`) on every single cycle when we try to consume data from the shard. If not locked, we need to stop consuming and probably need to bring the shard back to the lock manager for lease competition when the network comes back for all the instances.

@Siddharth-Somani ,

To bring your own shared lock registry there is just enough to declare the bean for it - and the one in the Kinesis binder is going to be overridden. But having an observation explained above, the custom `LockRegistry` won't help. Because the flaw not checking the lock on each polling cycle remains unchanged.

Unfortunately, I don't see how the number of shards may have an effect on the consuming logic.

Let's hope when we have a fix for the problem described above, it is going to fix your problem as well!

Any other thoughts?

**Comment by Siddharth-Somani** [comment]
@artembilan I am pretty sure about the number of shards having an effect on the consuming logic. We had never faced this issue when we had just 1 shard and multiple nodes. This issue happened on increasing the number of shards and that is when I reported this issue. Anyways, when is the fix planned to be released? Thanks for all the help until now.

**Comment by artembilan** [comment]
I'm looking into the solution right now. Hope to have something until the end of this week.
It would be great if you (both) can test your solution with my SNAPSHOT version before we really go for GA.

Thanks for understanding.

P.S. I still don't see any difference when we have two or more shards...

**Comment by malvine** [comment]
@artembilan happy to test yes.

@artembilan @Siddharth-Somani 
About the number of shards - I've seen the similar behaviour when scaling in/out the number of shards. If you scale in or out the shards - it's better to bounce your instances otherwise there is a high chance you'll get into the same situation as with network outage - multiple instances consuming from the shard.

Also, my colleague shared his example of duplicates as well: 
Let's take the simplest example again: one stream with one shard and one consumer with two instances.
One instance was elected as a lease owner, then they reached the read capacity for the DynamoDB Lock table. The second instance has been elected and now we end up with the multiple instances consuming the same shard.

But honestly saying I think all this is related to the lock issue.

**Comment by artembilan** [comment]
Those are good samples, @malvine !

I think I just underestimated the heartbeat of the lock. :shame: 
What I'm going to to as an implementation is just plain `tryLock()` when we are in a polling cycle. So, this way we would ensure that the lock is still our and it is renewed, of course. If anything fails we just go to the stop state of the shard consumer and push its shard to the consuming candidates for the next possible locking cycle.

**Comment by artembilan** [comment]
Please, find the fix here: https://github.com/spring-projects/spring-integration-aws/commit/ac74dfd2368c5c4b74793c259312313ad21ed5f8.

It would be great if you can try your solution with the latest `spring-integration-aws:2.3.5.BUILD-SNAPSHOT`.

Thank you for your patience!

**Comment by Siddharth-Somani** [comment]
@artembilan Have been getting these error logs after upgrading spring-integration-aws  - 
1. ```2020-12-19 15:47:23.515 ERROR 1 --- [s-shard-locks-1] a.i.k.KinesisMessageDrivenChannelAdapter : Error during locking: DynamoDbLock [lockKey=kinesis_stream_test-group:kinesis_stream_test:shardId-000000000035,lockedAt=2020-12-19@14:44:17.846, lockItem=LockItem{Partition Key=kinesis_stream_test-group:kinesis_stream_test:shardId-000000000035, Sort Key=Optional[SpringIntegrationLocks], Owner Name=appd-deployment-67f68d9d8c, Lookup Time=58635015, Lease Duration=20000, Record Version Number=702696d4-fca2-4ccb-a1f7-097b688cf6e7, Delete On Close=true, Is Released=false}]



org.springframework.dao.CannotAcquireLockException: Failed to lock at kinesis_stream_test-group:kinesis_stream_test:shardId-000000000035; nested exception is com.amazonaws.services.dynamodbv2.model.LockNotGrantedException: Cannot send heartbeat because lock is not granted

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.rethrowAsLockException(DynamoDbLockRegistry.java:406) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:522) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:485) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumerManager.run(KinesisMessageDrivenChannelAdapter.java:1495) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java) [na:1.8.0_242]

	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java) [na:1.8.0_242]

	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_242]

Caused by: com.amazonaws.services.dynamodbv2.model.LockNotGrantedException: Cannot send heartbeat because lock is not granted

	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBLockClient.sendHeartbeat(AmazonDynamoDBLockClient.java:1085) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBLockClient.sendHeartbeat(AmazonDynamoDBLockClient.java:1048) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at com.amazonaws.services.dynamodbv2.LockItem.sendHeartBeat(LockItem.java:254) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.doLock(DynamoDbLockRegistry.java:531) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:511) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]
```

2. ```2020-12-19 15:49:38.796  INFO 1 --- [is-dispatcher-1] a.i.k.KinesisMessageDrivenChannelAdapter : The lock for key 'kinesis_stream-group:kinesis_stream:shardId-000000000035' was not renewed in time



java.util.concurrent.TimeoutException: null

	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1784) ~[na:1.8.0_242]

	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928) ~[na:1.8.0_242]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumer.renewLockIfAny(KinesisMessageDrivenChannelAdapter.java:1031) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumer.execute(KinesisMessageDrivenChannelAdapter.java:943) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ConsumerDispatcher.run(KinesisMessageDrivenChannelAdapter.java:853) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java) [na:1.8.0_242]

	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_242]```

**Comment by artembilan** [comment]
@Siddharth-Somani ,

Thanks for testing this so quickly!

So, what I see that I have missed to fulfill the `Future`  when exception happens:
```
catch (Exception e) {
	logger.error("Error during locking: " + lock, e);
}
```

Will fix it soon.
On the other hand it is confirmed that there is something happening in your environment.
When we reach the code like this:
```
if (this.lockItem != null) {
	this.lockItem.sendHeartBeat();
	acquired = true;
}
```
this means that we have been here before and we could acquire the lock. 
What the problem is right now that when we try to re-lock, it doesn't belong us any more.
We end up with this logic:
```
if (lockItem.isExpired() || !lockItem.getOwnerName().equals(this.ownerName) || lockItem.isReleased()) {
           this.locks.remove(lockItem.getUniqueIdentifier());
            throw new LockNotGrantedException("Cannot send heartbeat because lock is not granted");
}
```
Therefore it is interesting to understand why your lock is expired and therefore it takes a new owner.
So, what is your `leaseDuration` and `heartbeatPeriod`?

**Comment by artembilan** [comment]
I have just pushed the fix: https://github.com/spring-projects/spring-integration-aws/commit/526353c9d80c6234e42fbf06ae84f81b77a4f7eb.

Give it 5-10 mins to make a fresh SNAPSHOT build.

Thanks for your help!

**Comment by Siddharth-Somani** [comment]
Hey thanks for the detailed explanation @artembilan . I have been using the default values itself, made no change. `leaseDuration` 20 seconds and `heartBeatPeriod` 5 seconds.

**Comment by artembilan** [comment]
So, it looks like you process your records in the target application much longer then those 20 seconds...
On the other hand by default there is that background thread for sending heartbeats every 5 seconds...

So, too many questions what is going on in your environment. And it is fully suspicious why that works for your with two shards.

Anyway, I'd be glad to hear from you back for the test against latest fix. Feel free to give me some answers to those my concerns as well!

**Comment by Siddharth-Somani** [comment]
Sure, I had reverted the changes. Let me upgrade it back to `spring-integration-aws:2.3.5.BUILD-SNAPSHOT`. Will let you know after I run the tests

**Comment by Siddharth-Somani** [comment]
I am still seeing the same error logs. Also, do you want me to update leaseDuration and heartBeatPeriod ?

**Comment by artembilan** [comment]
Yes, you still are going to see that `Error during locking: ` because the logic is like this:
```
catch (Exception e) {
	lockFuture.complete(false);
        logger.error("Error during locking: " + lock, e);
}
```
but you must not see ` was not renewed in time` any more. Because the logic is like this:
```
boolean lockRenewed = false;
try {
	lockRenewed = renewLockFuture.get(1, TimeUnit.SECONDS);
}
catch (Exception ex) {
	if (ex instanceof InterruptedException) {
		Thread.currentThread().interrupt();
	}
	logger.info("The lock for key '" + this.key + "' was not renewed in time", ex);
}
```
So, no `TimeoutException` means that we lost ownership correctly and the current instance must not consume shard any more.

**Comment by malvine** [comment]
@artembilan just played a bit with the latest `2.3.5.BUILD-SNAPSHOT`
For one stream with one shard and one consumer with two instances and network failure, that now looks promising.

if `instance one` holds a lock and network connection drops then I can see this
```
java.util.concurrent.TimeoutException: null
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1957) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2092) ~[na:na]
	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumer.renewLockIfAny(KinesisMessageDrivenChannelAdapter.java:1031) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar:na]
	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumer.execute(KinesisMessageDrivenChannelAdapter.java:943) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar:na]
	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ConsumerDispatcher.run(KinesisMessageDrivenChannelAdapter.java:853) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:830) ~[na:na]

```

Once the connection is restored, `instance one` picks up the lock and continues to get the messages. 
`Instance two` no longer gets the same messages as it used to with the `RELEASE` version.  Nice!



I will play a bit more with it and add also scaling up/down the shards etc. Will let you know.

**Comment by artembilan** [comment]
That's correct to see such a timeout exception since we really can't finish an execution because no network in between.
In the case of such an exception we just stop a consumer for this lock (and therefore shard) and schedule it for future attempts.
The `ShardConsumerManager` does it job then iterating candidates for `tryLock` until they are locked.

Thank you, @malvine , for sharing your experience and for testing this uneasy fix!

**Comment by malvine** [comment]
Hi @artembilan 
I played a bit more with `2.3.5.BUILD-SNAPSHOT` and I get an interesting INFO in the logs for `instance2` after scaling to 2 shards.
```
INFO a.i.k.KinesisMessageDrivenChannelAdapter : The [ShardConsumer{shardOffset=KinesisShardOffset{iteratorType=AFTER_SEQUENCE_NUMBER, sequenceNumber='49614242609631570993743825069427662177189212994282717186', timestamp=null, stream='test-stream', shard='shardId-000000000000', reset=false}, state=NEW}] has been started.
INFO  a.i.k.KinesisMessageDrivenChannelAdapter : Got an exception com.amazonaws.services.kinesis.model.InvalidArgumentException: Invalid SequenceNumber for ShardIteratorType AFTER_SEQUENCE_NUMBER, SequenceNumber has reached max possible value for the shard. (Service: AmazonKinesis; Status Code: 400; Error Code: InvalidArgumentException; Request ID: da14db58-6556-311e-8279-f41d4a5fe657; Proxy: null) during [ShardConsumer{shardOffset=KinesisShardOffset{iteratorType=AFTER_SEQUENCE_NUMBER, sequenceNumber='49614242609631570993743825069427662177189212994282717186', timestamp=null, stream='test-stream', shard='shardId-000000000000', reset=false}, state=NEW}] task invocation.
Process will be retried on the next iteration.
```

**How to replicate:**
I have one stream with one shard and two instances of the same consumer.
Initial setup we have `instance1` is consuming the `shardId-0`. `instance2` is not consuming anything. All good so far.

Then **I scale the stream to 2 shards**.
Now I got `instance2` consumes from `shardId-0`, `instance1` consumes from both `shardId-1` and `shardId-2`. Which is also fine and seems like everything is working - producer is able to produce messages and `instance1 ` is able to consume all of them without any loss. 
The only thing is `instance2` which is pointing to `shardId-0` has those INFO logs printed every second.
It stops only if I bounce `instance2`. Then those logs sto coming and `shardId-0`  lock key disappears from the `SpringIntegrationLockRegistry` DynamoDb table, and `shardId-0` stays  in `SpringIntegrationLockRegistry` table. `Instance1` continues to consume both `shardId-1` and `shardId-2` successfully.

**Comment by artembilan** [comment]
@malvine ,

thanks for sharing that, but this looks like a different problem and not related to our duplication consumption.
Please, consider to raise a new issue with more relevant details. 
We probably need to inverstigate how AWS Kinesis deals with problems like that when re-sharding happens.

From your other points it sounds like an original problem discussed here is really fixed.
I'll give @Siddharth-Somani a couple days to come back to us then I'll close this issue as fixed.

**Comment by malvine** [comment]
@artembilan agree.

I haven't seen any duplication issues with `2.3.5.BUILD-SNAPSHOT` so far. 👍

**Comment by Siddharth-Somani** [comment]
Hey @artembilan , I didn't get a chance to proceed further due to the year end holidays. I don't seem to get past the error logs that I shared above. With the older version `2.3.4`, my consumer is definitely not taking more than 20 seconds to process a record. Somehow with `2.3.5.BUILD-SNAPSHOT` , I am seeing the error that I mentioned above. Not sure on how I can proceed further.

**Comment by malvine** [comment]
@artembilan  I decided to re-run the timeout use- case the one where my consumer lost the connection to the stream against `2.3.5.BUILD-SNAPSHOT`. And it was working last year (I mean a couple of weeks ago). I am pretty sure.

But now I am seeing same error as @Siddharth-Somani  mentioned

This doesn't happen with one stream, one shard and one instance of the consumer. It manages to recover when the connection is restored.
But it happens with two instances of the same consumer and also with multiple shards as well. (no scaling involved at this point).

```
2020-12-19 15:47:23.515 ERROR 1 --- [s-shard-locks-1] a.i.k.KinesisMessageDrivenChannelAdapter : Error during locking: DynamoDbLock [lockKey=kinesis_stream_test-group:kinesis_stream_test:shardId-000000000035,lockedAt=2020-12-19@14:44:17.846, lockItem=LockItem{Partition Key=kinesis_stream_test-group:kinesis_stream_test:shardId-000000000035, Sort Key=Optional[SpringIntegrationLocks], Owner Name=appd-deployment-67f68d9d8c, Lookup Time=58635015, Lease Duration=20000, Record Version Number=702696d4-fca2-4ccb-a1f7-097b688cf6e7, Delete On Close=true, Is Released=false}]



org.springframework.dao.CannotAcquireLockException: Failed to lock at kinesis_stream_test-group:kinesis_stream_test:shardId-000000000035; nested exception is com.amazonaws.services.dynamodbv2.model.LockNotGrantedException: Cannot send heartbeat because lock is not granted

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.rethrowAsLockException(DynamoDbLockRegistry.java:406) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:522) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:485) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumerManager.run(KinesisMessageDrivenChannelAdapter.java:1495) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java) [na:1.8.0_242]

	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java) [na:1.8.0_242]

	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_242]

Caused by: com.amazonaws.services.dynamodbv2.model.LockNotGrantedException: Cannot send heartbeat because lock is not granted

	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBLockClient.sendHeartbeat(AmazonDynamoDBLockClient.java:1085) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBLockClient.sendHeartbeat(AmazonDynamoDBLockClient.java:1048) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at com.amazonaws.services.dynamodbv2.LockItem.sendHeartBeat(LockItem.java:254) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.doLock(DynamoDbLockRegistry.java:531) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:511) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

```


**The easiest use case:**
one stream with one shard and one consumer with two instances.

`instance 1` has a lease.
internet connection drops. logs suggest we lost connection to dynamoDb.
internet connection restored. `instance 2` has a lease and consuming the stream. `instance 1` fires some ERRORs as above every 2 seconds. 
Even if I shutdown the `instance2` (the one that has a lease now and consumes the stream), the `instance1` is still not able to recover fires those ERROR logs and doesn't consume the stream. Only restart of the instance brings it back to life.

**Comment by artembilan** [comment]
>  Cannot send heartbeat because lock is not granted

Probably the problem that we send a heartbeat from the `DynamoDbLock`, but don't react to the error from its outcome.
I'll take a look into that later this week.

Thanks for reporting!

**Comment by artembilan** [comment]
I have just fixed heart-beat logic in the `DynamoDbLockRegistry`, so would be great if you can re-test your solution with the latest `2.3.5.BUILD-SNAPSHOT`.

Thanks

**Comment by Siddharth-Somani** [comment]
@artembilan I am still seeing the 2 error logs that I shared above in a multiple instance environment. @malvine Can you also test it out by running multiple instances ?

**Comment by artembilan** [comment]
Please , share those fresh logs : I’m lost already I so long discussion with several topics. Also: any chances to heat from you back a confirmation that we don’t have duplicates any more?

**Comment by Siddharth-Somani** [comment]
@artembilan Due to the error, not all messages are being consumed. Only a few of them are being consumed and the remaining messages remain in kinesis itself. However, in those few messages that are consumed, no duplicates are seen.

Logs - 
1. 
```2021-01-07 18:51:27.589 ERROR 1 --- [s-shard-locks-1] a.i.k.KinesisMessageDrivenChannelAdapter : Error during locking: DynamoDbLock [lockKey=kinesis_stream_test-group:kinesis_stream_test:shardId-000000000037,lockedAt=2021-01-07@13:21:24.840, lockItem=LockItem{Partition Key=kinesis_stream_test-group:kinesis_stream_test:shardId-000000000037, Sort Key=Optional[SpringIntegrationLocks], Owner Name=appd-deployment-76cd6cc56d, Lookup Time=1440510766, Lease Duration=20000, Record Version Number=fafa8d64-506e-4c9d-af8c-a52fac03657a, Delete On Close=true, Is Released=false}]



org.springframework.dao.CannotAcquireLockException: Failed to lock at kinesis_stream_test-group:kinesis_stream_test:shardId-000000000037; nested exception is com.amazonaws.services.dynamodbv2.model.LockNotGrantedException: Cannot send heartbeat because lock is not granted

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.rethrowAsLockException(DynamoDbLockRegistry.java:406) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:522) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:485) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumerManager.lambda$run$0(KinesisMessageDrivenChannelAdapter.java:1450) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at java.util.Collection.removeIf(Collection.java:414) ~[na:1.8.0_242]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumerManager.run(KinesisMessageDrivenChannelAdapter.java:1443) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java) ~[na:1.8.0_242]

	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java) ~[na:1.8.0_242]

	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_242]

Caused by: com.amazonaws.services.dynamodbv2.model.LockNotGrantedException: Cannot send heartbeat because lock is not granted

	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBLockClient.sendHeartbeat(AmazonDynamoDBLockClient.java:1085) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at com.amazonaws.services.dynamodbv2.AmazonDynamoDBLockClient.sendHeartbeat(AmazonDynamoDBLockClient.java:1048) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at com.amazonaws.services.dynamodbv2.LockItem.sendHeartBeat(LockItem.java:254) ~[dynamodb-lock-client-1.1.0.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.doLock(DynamoDbLockRegistry.java:531) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.lock.DynamoDbLockRegistry$DynamoDbLock.tryLock(DynamoDbLockRegistry.java:511) ~[spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	... 9 common frames omitted```
	
2. 
```2021-01-07 18:53:49.299  INFO 1 --- [is-dispatcher-1] a.i.k.KinesisMessageDrivenChannelAdapter : The lock for key 'kinesis_stream_test-group:kinesis_stream_test:shardId-000000000037' was not renewed in time



java.util.concurrent.TimeoutException: null

	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1784) ~[na:1.8.0_242]

	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928) ~[na:1.8.0_242]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumer.renewLockIfAny(KinesisMessageDrivenChannelAdapter.java:1031) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumer.execute(KinesisMessageDrivenChannelAdapter.java:943) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ConsumerDispatcher.run(KinesisMessageDrivenChannelAdapter.java:853) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java) [na:1.8.0_242]

	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_242]```

**Comment by artembilan** [comment]
@Siddharth-Somani , the 
```
Caused by: com.amazonaws.services.dynamodbv2.model.LockNotGrantedException: Cannot send heartbeat because lock is not granted
```
was fixed in the latest commit: https://github.com/spring-projects/spring-integration-aws/commit/696472037fd841c5e4c80f361ddc12425f22a09e

So, looks like you are reporting out-dated info.
Please, consider to check your solution with the latest `2.3.5.BUILD-SNAPSHOT` - `-U` for Maven to refresh your local snapshots.

**Comment by Siddharth-Somani** [comment]
My bad @artembilan , will check and get back to you

**Comment by malvine** [comment]
@artembilan  @Siddharth-Somani  

with the latest `2.3.5.BUILD-SNAPSHOT` taken today:

**Test case 1**: **_Test with one stream and one shard and multiple instances of the consumer_.**
**Outcome**: ✅ **All good**. If the connection is dropped for all the instances - I can see exceptions fired on the instance that had an initial lease. Once the connection is restored whatever gets the lease first continues to consume the shard - exceptions are no longer thrown on any of the instances. And no duplicate records.


**Test case 2**: **_Test with one stream and two shards and multiple instances of the consumer_**.
**Outcome**: ✅ Same behaviour observed similar to the test above. 
_(note: two shards have been created initially during the stream creation - not but scaling out - there are still some problems with scaling the stream, but it's not related to this issue at all.)_

**Test case 3**: ** Test with one stream and two shards and one instance of the consumer_**.
**Outcome**: ✅  Exceptions are thrown on the regular basis in the instance logs, but once the connection is restored the exceptions are no longer thrown and lease assigned to the consumer instance and records are being consumed.

**Comment by artembilan** [comment]
Sounds good, @malvine .
Thank you for feedback!

@Siddharth-Somani , we just wait for you to confirm.

After that we probably are good to go with a `2.3.5.RELEASE`.

**Comment by Siddharth-Somani** [comment]
@artembilan Give me a day's time. Will update! Thanks

**Comment by sabbyanandan** [comment]
@Siddharth-Somani @malvine: Thank you for continuing to push and help test the SNAPSHOT bits.

**Comment by Siddharth-Somani** [comment]
@artembilan Will it be okay to delete the entries in `SpringIntegrationLockRegistry` and `SpringIntegrationMetadataStore` so that the new instance values and shard ids would be auto populated? . I am having a few issues in testing and hence would want to start a fresh and update here ASAP.

**Comment by artembilan** [comment]
Since we talk about testing this particular issue, it is probably better to start from a clean DB.
If there are some problems with existing entries in DB, then it is probably not related to this one.
Although the existing checkpoint should not effect the behavior. It is really expected logic to be able to obtain a shard in a new consumer and continue consuming from where it is left by some other consumer in that checkpoint store.

The lock registry should auto-clean itself since there is really a lease duration and if the lock record was not removed somehow by unlocking functionality, then it must be expired the next time we try to lock it if its lease duration is exceeded.

Of course I'm talking about the perfect solution which we have thought about when we designed all of this components.
But as you see in practice we always missing something and it is really better to test the fix from empty DB to make it as clean as possible without any side effects which may impact the final report. If you confirm the fix with a simple testing scenario, we always can go ahead with more complicated one afterwards.

**Comment by Siddharth-Somani** [comment]
@artembilan Sorry to keep this thread waiting. I would like to state an observation from a test I did at scale today ( 3 consumer instances, 8 kinesis shards, 70 TPS (throughput per second to kinesis). 

The test ran successfully for the first 15 mins without any issues. After that the following log across all the shards started to appear and the consumption of records stopped -
```
2021-01-13 18:19:05.222  INFO 1 --- [is-dispatcher-1] a.i.k.KinesisMessageDrivenChannelAdapter : The lock for key 'data_stream-group:data_stream:shardId-000000000057' was not renewed in time



java.util.concurrent.TimeoutException: null

	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1784) ~[na:1.8.0_242]

	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928) ~[na:1.8.0_242]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumer.renewLockIfAny(KinesisMessageDrivenChannelAdapter.java:1031) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ShardConsumer.execute(KinesisMessageDrivenChannelAdapter.java:943) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at org.springframework.integration.aws.inbound.kinesis.KinesisMessageDrivenChannelAdapter$ConsumerDispatcher.run(KinesisMessageDrivenChannelAdapter.java:853) [spring-integration-aws-2.3.5.BUILD-SNAPSHOT.jar!/:na]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_242]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_242]

	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_242]
	``` 

The above log did come when the test was working as expected but not continuously. The only way to resume consumption of records was to kill old instances and bring up new instances. And again stops consumption after 5-10 mins of run with the same error log.

**Comment by artembilan** [comment]
@Siddharth-Somani ,

thanks for coming back to us!

So, we fail here:
```
try {
	lockRenewed = renewLockFuture.get(1, TimeUnit.SECONDS);
}
catch (Exception ex) {
	if (ex instanceof InterruptedException) {
		Thread.currentThread().interrupt();
	}
	logger.info("The lock for key '" + this.key + "' was not renewed in time", ex);
}
```
Kinda that `1 second` is not enough in your environment to get an answer from the:
```
if (lock.tryLock()) {
	try {
		lockFuture.complete(true);
	}
	finally {
		lock.unlock();
	}
}
else {
	lockFuture.complete(false);
	this.locks.remove(lockFuture.lockKey);
}
```

I'm probably not right and it is better to have this timeout as `10 seconds` to give AWS bigger chance to finish all the requests.

Are you sure that there no more any logs around that problem?

**Comment by Siddharth-Somani** [comment]
Yes, 1 second does seem less. No more logs ( I haven't turned on the DEBUG logs. Can do so if you need me to ).

**Comment by artembilan** [comment]
OK. Thanks for confirmation!

My concern is to be sure how long an AWS request is processed and that DynamoDB Lock client doesn't fail somehow suspiciously, e.g. with the `ProvisionedThroughputExceededException` or so.

But anyway: I will look into some property exposure for this kind of timeouts in the `KinesisMessageDrivenChannelAdapter`.

**Comment by artembilan** [comment]
See latest commit: https://github.com/spring-projects/spring-integration-aws/commit/ea135917f3da8efc51102755c58b24d21a30fae2.

I have added a `lockRenewalTimeout` property with the default as `10 seconds`.

Let's hop it helps somehow with your environment!

**Comment by Siddharth-Somani** [comment]
Hey @artembilan , ran another test with the latest fix. 
1. Good news is that - I don't see duplicate messages being consumed by multiple hosts. I think this problem is resolved. 
2. Coming to the issues - I still see the `lock not renewed in time` log coming. An observation on this log is made in the next point.
3. The consumption of records is extremely slow. I am currently running with a 3 instances and 8 kinesis shards setup and here are a few observations based on the consumption rate image below -
<img width="1680" alt="Screenshot 1399-10-26 at 12 02 33 AM" src="https://user-images.githubusercontent.com/18643496/104633380-0eb1a000-56c5-11eb-9b52-92ea6dc8ba57.png">

- Between 10:40 pm to 11:25 pm, 2 consumer instances were consuming the records and in the dynamo table `SpringIntegrationLockRegistry` , these 2 instances were divided among the 8 shards. As you see, the consumption is extremely slow and I also saw the `lock not renewed in time` log across all the shards

- Between 11:25 pm and 11:50 pm, the consumption of records suddenly stopped and I could still see the `lock not renewed in time` log across all the shards

- I restarted the instances and in the `SpringIntegrationLockRegistry` table, all shards were assigned to just one instance. And then from 11:50 pm to 12 am, the consumption speed was good and I did not see the `lock not renewed in time` log (in the setup where one instance is shared across all shards in the dynamo table)

**Comment by artembilan** [comment]
Ok. I think we can go with this so far and take a look into slow behavior a bit later.

There is that loop over the lock we are waiting for and probably the `tryLock()` operation on DynamoDB is not fast enough.
Plus we have there `sleep(250)` to avoid busy-spin and rate limit problem.

Not sure how to fix all of that, but it must be as a part of separate issue.
Feel free to raise one in the https://github.com/spring-projects/spring-integration-aws project!

Thank you for great contribution! Even if not code, just reasonable discussion and testing is already a good contribution into this open source project.

I think we will go release next...

**Comment by Siddharth-Somani** [comment]
Sounds good @artembilan , I will raise a new issue once the snapshot is released. Thanks for all the help.

**Comment by artembilan** [comment]
Fixed in the https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/releases/tag/v2.0.4.RELEASE and https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/releases/tag/v2.1.0
